File: LeetCode/Explore/Queue & Stack/287. Find the Duplicate Number.py
#Time: O(n)
#Space: O(1)
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        if not nums:
            return 0
        
        slow = nums[0]
        fast = nums[0]
        start = nums[0]
        
        while True:
            slow = nums[slow]
            fast = nums[nums[fast]]
            if slow == fast:
                break
 
        while start != slow:
            slow = nums[slow]
            start = nums[start]
        
        return start

File: LeetCode/Problems/C++/287. Find the Duplicate Number.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        if(!nums.size()) return -1;
        
        int fast = nums[0];
        int slow = nums[0];
        
        while(true){
            fast = nums[nums[fast]];
            slow = nums[slow];
            if(fast == slow) break;
        }
        
        slow = nums[0];
        
        while(slow!=fast){
            slow = nums[slow];
            fast = nums[fast];
        }
        return fast;
    }
    
};